### PR TITLE
US122994: Implement min-exclusive and max-exclusive

### PR DESCRIPTION
**Pull request by krshnpatel-d2l**
## [US122994](https://rally1.rallydev.com/#/15545167705d/custom/21568985922?detail=%2Fuserstory%2F460347867248&fdp=true)

~ Implemented `min-exclusive` and `max-exclusive` attributes
~ Added a demo item that uses both attributes
~ Updated validation lang terms for input number
~ Made sure native input validation still works as expected (due to changes to lang term)
~ Updated unit tests for input number
~ Updated input-number docs

### Unrelated Changes
~ Refactored unit tests for input number
~ Updated localize-mixin readme

### CODE REVIEW

File: components/inputs/test/input-number.test.js [hunk 1]
```diff
@@ -61,44 +64,50 @@ describe('d2l-input-number', () => {
...
+				expectInnerValue: '1.234'
+			}
+		].forEach((test) => {
+			it(test.name, async() => {
+				const elem = await fixture(test.fixture);
+				elem.value = test.value;
+				await elem.updateComplete;
+				expect(elem.value).to.equal(test.expectedValue);
+				expect(getInnerInputValue(elem)).to.equal(test.expectInnerValue);
+			});
```
Comment (krshnpatel-d2l) [inline components/inputs/test/input-number.test.js:110]: These tests haven't changed, just refactored.

File: components/inputs/test/input-number.test.js [hunk 2]
```diff
@@ -128,71 +137,91 @@ describe('d2l-input-number', () => {
...
+				fixture: minExclusiveFixture,
+				value: 5,
+				expectedError: 'Number must be higher than 5.'
+			},
+			{
+				name: 'should be invalid if number is equal to max',
+				fixture: maxExclusiveFixture,
+				value: 10,
+				expectedError: 'Number must be lower than 10.'
+			}
```
Comment (krshnpatel-d2l) [inline components/inputs/test/input-number.test.js:215]: These are the new tests I added.

File: components/inputs/test/input-number.test.js [hunk 3]
```diff
@@ -128,71 +137,91 @@ describe('d2l-input-number', () => {
...
+				name: 'should be valid if number is in range',
+				fixture: minMaxFixture,
+				value: 7,
+				expectedError: ''
+			},
+			{
+				name: 'should be invalid if number is out of range',
+				fixture: minMaxFixture,
+				value: 1,
+				expectedError: 'Number must be higher than or equal to 5 and lower than or equal to 10.'
```
Comment (krshnpatel-d2l) [inline components/inputs/test/input-number.test.js:166]: Expected error now includes "or equal to".

File: components/inputs/test/input-number.test.js [hunk 4]
```diff
@@ -128,71 +137,91 @@ describe('d2l-input-number', () => {
...
+				name: 'should be valid if number is higher than min',
+				fixture: minFixture,
+				value: 10,
+				expectedError: ''
+			},
+			{
+				name: 'should be invalid if number is lower than min',
+				fixture: minFixture,
+				value: 1,
+				expectedError: 'Number must be higher than or equal to 5.'
```
Comment (krshnpatel-d2l) [inline components/inputs/test/input-number.test.js:178]: Expected error now includes "or equal to".

File: components/inputs/test/input-number.test.js [hunk 5]
```diff
@@ -128,71 +137,91 @@ describe('d2l-input-number', () => {
...
+				name: 'should be valid if number is lower than max',
+				fixture: maxFixture,
+				value: 1,
+				expectedError: ''
+			},
+			{
+				name: 'should be invalid if number is higher than max',
+				fixture: maxFixture,
+				value: 15,
+				expectedError: 'Number must be lower than or equal to 10.'
```
Comment (krshnpatel-d2l) [inline components/inputs/test/input-number.test.js:190]: Expected error now includes "or equal to".

File: lang/en.js [hunk 6]
```diff
@@ -10,9 +10,9 @@ export default {
 	"components.form-element.defaultError": "{label} is invalid.",
 	"components.form-element.defaultFieldLabel": "Field",
 	"components.form-element.input.email.typeMismatch": "Email is not valid",
-	"components.form-element.input.number.rangeError": "Number must be between {min} and {max}.",
-	"components.form-element.input.number.rangeOverflow": "Number must be lower than {max}.",
-	"components.form-element.input.number.rangeUnderflow": "Number must be higher than {min}.",
+	"components.form-element.input.number.rangeError": "Number must be higher than{minInclusive} {min} and lower than{maxInclusive} {max}.",
+	"components.form-element.input.number.rangeOverflow": "Number must be lower than{maxInclusive} {max}.",
+	"components.form-element.input.number.rangeUnderflow": "Number must be higher than{minInclusive} {min}.",
```
[the remark this answers is not in the document]
Comment (krshnpatel-d2l) [reply lang/en.js:15]: Ahhh, yes, this is true. I will change it up in the next commit. 👍

File: components/inputs/input-number.js [hunk 7]
```diff
@@ -58,6 +60,8 @@ class InputNumber extends SkeletonMixin(FormElementMixin(LocalizeCoreElement(Lit
 		this.disabled = false;
 		this.inputWidth = '4rem';
 		this.labelHidden = false;
+		this.maxExclusive = false;
+		this.minExclusive = false;
```
Comment (dlockhart) [inline components/inputs/input-number.js:64]: README list of attributes will need updating.

File: lang/en.js [hunk 8]
```diff
@@ -10,9 +10,9 @@ export default {
 	"components.form-element.defaultError": "{label} is invalid.",
 	"components.form-element.defaultFieldLabel": "Field",
 	"components.form-element.input.email.typeMismatch": "Email is not valid",
-	"components.form-element.input.number.rangeError": "Number must be higher than{minInclusive} {min} and lower than{maxInclusive} {max}.",
-	"components.form-element.input.number.rangeOverflow": "Number must be lower than{maxInclusive} {max}.",
-	"components.form-element.input.number.rangeUnderflow": "Number must be higher than{minInclusive} {min}.",
+	"components.form-element.input.number.rangeError": "Number must be {minExclusive, select, true {greater than} false {greater than or equal to}} {min} and {maxExclusive, select, true {less than} false {less than or equal to}} {max}.",
+	"components.form-element.input.number.rangeOverflow": "Number must be {maxExclusive, select, true {less than} false {less than or equal to}} {max}.",
+	"components.form-element.input.number.rangeUnderflow": "Number must be {minExclusive, select, true {greater than} false {greater than or equal to}} {min}.",
```
Comment (dlockhart) [inline lang/en.js:15]: We'll need to be extra careful when the translations for this come back to ensure they get translated properly. Can you double-check that for a different language which doesn't have the extra replacements that it at least doesn't blow up by passing in the minExclusive/maxExclusive value?

**Review by dlockhart — APPROVED** [review]

File: lang/en.js (hunk 8 above)
Comment (krshnpatel-d2l) [reply lang/en.js:15]: Tested it with `fr` and nothing seems to blow up

**Pull request merged**